From 39209d72fbaad0921b97a171d8c503e76d2b994e Mon Sep 17 00:00:00 2001
From: Nikolay Aleksandrov <nikolay@redhat.com>
Date: Fri, 5 Sep 2014 11:04:03 -0400
Subject: [ethernet] sfc: Add 40G link capability decoding

Message-id: <1409915048-26438-28-git-send-email-nikolay@redhat.com>
Patchwork-id: 91427
O-Subject: [RHEL7.1 PATCH BZ 1110888 27/32] sfc: Add 40G link capability decoding
Bugzilla: 1110888
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Edward Cree <ecree@solarflare.com>

Needed to select 40G mode on a 10G/40G capable card.

Signed-off-by: Shradha Shah <sshah@solarflare.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit ac331e948346f2706cd82b2d259800621b9db04f)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/sfc/mcdi_port.c b/drivers/net/ethernet/sfc/mcdi_port.c
index 91d2325..3361926 100644
--- a/drivers/net/ethernet/sfc/mcdi_port.c
+++ b/drivers/net/ethernet/sfc/mcdi_port.c
@@ -183,6 +183,8 @@ static u32 mcdi_to_ethtool_cap(u32 media, u32 cap)
    result |= SUPPORTED_1000baseKX_Full;
   if (cap & (1 << MC_CMD_PHY_CAP_10000FDX_LBN))
    result |= SUPPORTED_10000baseKX4_Full;
+  if (cap & (1 << MC_CMD_PHY_CAP_40000FDX_LBN))
+   result |= SUPPORTED_40000baseKR4_Full;
   break;
 
  case MC_CMD_MEDIA_XFP:
@@ -190,6 +192,12 @@ static u32 mcdi_to_ethtool_cap(u32 media, u32 cap)
   result |= SUPPORTED_FIBRE;
   break;
 
+ case MC_CMD_MEDIA_QSFP_PLUS:
+  result |= SUPPORTED_FIBRE;
+  if (cap & (1 << MC_CMD_PHY_CAP_40000FDX_LBN))
+   result |= SUPPORTED_40000baseCR4_Full;
+  break;
+
  case MC_CMD_MEDIA_BASE_T:
   result |= SUPPORTED_TP;
   if (cap & (1 << MC_CMD_PHY_CAP_10HDX_LBN))
@@ -237,6 +245,8 @@ static u32 ethtool_to_mcdi_cap(u32 cap)
   result |= (1 << MC_CMD_PHY_CAP_1000FDX_LBN);
  if (cap & (SUPPORTED_10000baseT_Full | SUPPORTED_10000baseKX4_Full))
   result |= (1 << MC_CMD_PHY_CAP_10000FDX_LBN);
+ if (cap & (SUPPORTED_40000baseCR4_Full | SUPPORTED_40000baseKR4_Full))
+  result |= (1 << MC_CMD_PHY_CAP_40000FDX_LBN);
  if (cap & SUPPORTED_Pause)
   result |= (1 << MC_CMD_PHY_CAP_PAUSE_LBN);
  if (cap & SUPPORTED_Asym_Pause)
@@ -285,6 +295,7 @@ static u32 mcdi_to_ethtool_media(u32 media)
 
  case MC_CMD_MEDIA_XFP:
  case MC_CMD_MEDIA_SFP_PLUS:
+ case MC_CMD_MEDIA_QSFP_PLUS:
   return PORT_FIBRE;
 
  case MC_CMD_MEDIA_BASE_T:
-- 
1.7.1